Enable setting the TypeMap entrypoint assembly via a RuntimeHostConfigurationOption

jtschuster opened the pull request:
Implements the alternative to https://github.com/dotnet/runtime/issues/121138 mentioned in https://github.com/dotnet/runtime/issues/121138#issuecomment-3459953860

Enables the TypeMap to be used in Library output types. Projects can set the following RuntimeHostConfigurationOption to enable this behavior. If it exists, this option takes priority over the entrypoint assembly.

```
<ItemGroup>
    <RuntimeHostConfigurationOption Include="System.Runtime.InteropServices.TypeMappingEntryAssembly"
        Value="$(TypeMapEntryAssembly)"
        Condition="$(TypeMapEntryAssembly) != ''" />
</ItemGroup>
```

jtschuster committed:
Enable TypeMap apis to use an RuntimeHostConfigurationOption to find the
entrypoint assembly for building the typemap.

jtschuster committed:
Missed a place where Main was renamed to Run

jtschuster committed:
Fix build break, generate runtimeconfig.json

sbomer commented on src/coreclr/tools/aot/ILCompiler/Program.cs:
```diff
@@ -314,8 +314,25 @@ public int Run()
+                foreach (var runtimeKnob in runtimeKnobs)
                 {
+                    var knobAndValue = runtimeKnob.Split('=', 2);
+                    if (knobAndValue.Length == 2 && knobAndValue[0] == "System.Runtime.InteropServices.TypeMappingEntryAssembly")
```
Nit: could this be "System.Runtime.InteropServices.TypeMapEntryAssembly" to match TypeMapAssemblyTargetAttribute?

jtschuster committed:
Make test Exe

jtschuster committed:
Need to set RequiresProcessIsolation for RuntimeConfig options?

jtschuster committed:
Use runtimeknob on NativeAOT

jtschuster committed:
Pass runtimeknobs to TrimmingDriver and fix test failure.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler/Program.cs:
```diff
@@ -314,8 +314,25 @@ public int Run()
+                foreach (var runtimeKnob in runtimeKnobs)
                 {
+                    var knobAndValue = runtimeKnob.Split('=', 2);
+                    if (knobAndValue.Length == 2 && knobAndValue[0] == "System.Runtime.InteropServices.TypeMappingEntryAssembly")
```
Would it be cleaner to introduce a first class command line switch for this populated from the `TypeMapEntryAssembly` property (the approach in https://github.com/dotnet/runtime/issues/121138#issuecomment-3459953860)

jtschuster replied:
Added a proper switch to the tools. We'll need to add something to the SDK to set the RuntimeHostConfigurationOption when TypeMapEntryAssembly is set or require projects to set both, otherwise we only would set it when PublishAOT or PublishTrimmed are set.

jtschuster committed:
Make TypeMapEntryAssembly a first-class switch in trim tools

However, it's still necessary to set the RuntimeConfigurationOption for
runtime behavior on CoreCLR.

jtschuster committed:
Forward --typemap-entry-assembly in ilc tests

MichalStrehovsky commented on src/tools/illink/src/linker/Linker/LinkContext.cs:
```diff
@@ -135,6 +137,8 @@ public Pipeline Pipeline
 
         public Dictionary<string, bool> FeatureSettings { get; init; }
 
+        public Dictionary<string, string> Features { get; init; }
```
Do we still need this?

MichalStrehovsky reviewed (approved):
Looks good otherwise!

jtschuster committed:
Remove Features property, fix alignment in trimmer help message

jtschuster committed:
Add compat suppressions

jtschuster committed:
Don't set RuntimeHostConfigurationOption.Trim

jtschuster committed:
Create instances of source types in the proxy type map for nativeaot

MichalStrehovsky commented on src/tests/Interop/TypeMap/TypeMapEntryAssemblyApp.cs:
```diff
@@ -0,0 +1,58 @@
+
+        // For NativeAOT, instantiate the type to ensure the mappings are marked
+        _ = new Lib5Type1();
+        _ = new Lib5Type2();
```
This should be rewritten to follow the existing pattern. Any runs that execute ILLinker would also be broken, this is not just for native AOT sake:

https://github.com/dotnet/runtime/blob/891c183b22d023eea7bc4aa57dbc219d54852036/src/tests/Interop/TypeMap/TypeMapApp.cs#L148-L149

jtschuster committed:
Use new().GetType() pattern in ProxyTypeMap test
